eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,231 @@
+      val introductionPayload = NodePayload(blindedHop.route.introductionNodeId, OutgoingBlindedPerHopPayload.createIntroductionPayload(blindedHop.route.introductionNode.encryptedPayload, blindedHop.route.introductionNode.blindingEphemeralKey))
+      val intermediateBlindedPayloads = blindedHop.route.subsequentNodes.dropRight(1).map(n => NodePayload(n.blindedPublicKey, OutgoingBlindedPerHopPayload.createIntermediatePayload(n.encryptedPayload)))
+      val finalBlindedPayload = NodePayload(blindedHop.route.blindedNodes.last.blindedPublicKey, OutgoingBlindedPerHopPayload.createFinalPayload(route.amount, totalAmount, expiry, blindedHop.route.blindedNodes.last.encryptedPayload, customTlvs))
+      val introductionAmount = route.amount + blindedHop.paymentInfo.fee(route.amount)
+      val introductionExpiry = expiry + blindedHop.paymentInfo.cltvExpiryDelta
+      val blindedPayloads = PaymentPayloads(introductionAmount, introductionExpiry, introductionPayload +: intermediateBlindedPayloads :+ finalBlindedPayload)
+      if (channelHops.isEmpty) {
+        // We are the introduction node of the blinded route.
+        blindedPayloads
+      } else {
+        channelHops.tail.reverse.foldLeft(blindedPayloads) {
+          case (current, channelHop) =>
+            val payload = NodePayload(channelHop.nodeId, IntermediatePayload.ChannelRelay.Standard(channelHop.shortChannelId, current.amount, current.expiry))
+            PaymentPayloads(current.amount + channelHop.fee(current.amount), current.expiry + channelHop.cltvExpiryDelta, payload +: current.payloads)
+        }
+      }
+    }
+  }
+
+  private def extractRoute(route: Route): Try[(Seq[ChannelHop], BlindedHop)] = {
+    // The payment route must start with channel hops to reach the introduction node of a blinded route.
+    val channelHops = route.hops.dropRight(1).collect { case hop: ChannelHop => hop }
+    val blindedHop_opt = route.hops.takeRight(1).collectFirst { case hop: BlindedHop => hop }
+    blindedHop_opt match {
+      case Some(blindedHop) if channelHops.length == route.hops.length - 1 => Success((channelHops, blindedHop))
+      case _ => Failure(new IllegalArgumentException(s"cannot send payment: invalid blinded route (${route.printChannels()})"))
+    }
+  }
+}
+
+object BlindedRecipient {
+  def apply(invoice: Bolt12Invoice, totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Seq[GenericTlv]): BlindedRecipient = {
+    val extraEdges = invoice.blindedPaths.zip(invoice.blindedPathsInfo).map {
+      case (route, info) =>
+        // We don't know the scids of channels inside the blinded route, but it's useful to have an ID to refer to a given
+        // edge in the graph, so we create a dummy one for the duration of the payment attempt.
+        val dummyId = ShortChannelId.generateLocalAlias()
+        Invoice.BlindedEdge(dummyId, invoice.nodeId, route, info)
+    }
+    BlindedRecipient(invoice.nodeId, invoice.features, totalAmount, expiry, extraEdges, customTlvs)
```

Comment:
The `nodeId` should be generated randomly to ensure there is no public node with that id.

Reply:
That's a good question, I'm not sure yet. If we get an invoice with a `nodeId` that matches a node in the public graph, why couldn't we just pay it without using route blinding if we're able to find a normal route? If the recipient doesn't want that to happen, shouldn't it be their responsibility to generate a random `nodeId`?

It would be nice to keep the real `nodeId` we were given in the `Recipient` to get better accounting of who we paid. Also, the code currently doesn't support making the payment if we found a normal route instead of using the blinded hops, but that should be very simple to add (if we think it's a good idea). What do you think?

Reply:
We must use the blinded route, that was the whole point of making it mandatory. If the recipient doesn't want privacy, it can always use a blinded route with no hop.
And blinded route doesn't just provide privacy, it also has encrypted data that the recipient may need to finalize the payment.

Reply:
Fixed in the latest version.